src/iostream.c

```diff
@@ -531,15 +551,23 @@ Obj FuncREAD_IOSTREAM( Obj self, Obj stream, Obj len )
   Int ret;
   Obj string;
   string = NEW_STRING(INT_INTOBJ(len));
-  while (!PtyIOStreams[pty].inuse)
-    pty = INT_INTOBJ(ErrorReturnObj("IOSTREAM %d is not in use",pty,0L,
-                                    "you can replace stream number <num> via 'return <num>;'"));
+  HashLock(PtyIOStreams);
+  if (!PtyIOStreams[pty].inuse) {
+    HashUnlock(PtyIOStreams);
+    ErrorMayQuit("IOSTREAM %d is not in use",pty,0L);
+    return Fail;
+  }
+  
   /* HandleChildStatusChanges(pty);   Omit this to allow picking up "trailing" bytes*/
   ret = ReadFromPty2(pty, CSTR_STRING(string), INT_INTOBJ(len), 1);
-  if (ret == -1)
+  if (ret == -1) {
+    HashUnlock(PtyIOStreams);
     return Fail;
+  }
   SET_LEN_STRING(string, ret);
```

Comment:
Well actually, you mustn't call anything which might take a lock of any kind while a `HashLock` is held, including another `HashLock` -- but that might not be documented.